tfhe/src/high_level_api/integers/unsigned/base.rs

```diff
@@ -236,6 +236,230 @@ where
+    ///
+    /// let result = a.trailing_zeros();
+    /// let decrypted: u16 = result.decrypt(&client_key);
+    /// assert_eq!(decrypted, 3);
+    /// ```
+    pub fn trailing_zeros(&self) -> super::FheUint32 {
+        global_state::with_internal_keys(|key| match key {
+            InternalServerKey::Cpu(cpu_key) => {
+                let result = cpu_key
+                    .pbs_key()
+                    .trailing_zeros_parallelized(&*self.ciphertext.on_cpu());
+                let result = cpu_key.pbs_key().cast_to_unsigned(
+                    result,
+                    super::FheUint32Id::num_blocks(cpu_key.pbs_key().message_modulus()),
+                );
+                super::FheUint32::new(result)
+            }
+            #[cfg(feature = "gpu")]
+            InternalServerKey::Cuda(_) => {
+                panic!("Cuda devices do not support trailing_zeros yet");
+            }
+        })
+    }
+
+    /// Returns the number of trailing ones in the binary representation of self.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tfhe::prelude::*;
+    /// use tfhe::{generate_keys, set_server_key, ConfigBuilder, FheBool, FheUint16};
+    ///
+    /// let (client_key, server_key) = generate_keys(ConfigBuilder::default());
+    /// set_server_key(server_key);
+    ///
+    /// let a = FheUint16::encrypt(0b0000000_0110111u16, &client_key);
+    ///
+    /// let result = a.trailing_ones();
+    /// let decrypted: u16 = result.decrypt(&client_key);
+    /// assert_eq!(decrypted, 3);
+    /// ```
+    pub fn trailing_ones(&self) -> super::FheUint32 {
+        global_state::with_internal_keys(|key| match key {
+            InternalServerKey::Cpu(cpu_key) => {
+                let result = cpu_key
+                    .pbs_key()
+                    .trailing_ones_parallelized(&*self.ciphertext.on_cpu());
+                let result = cpu_key.pbs_key().cast_to_unsigned(
+                    result,
+                    super::FheUint32Id::num_blocks(cpu_key.pbs_key().message_modulus()),
+                );
+                super::FheUint32::new(result)
+            }
+            #[cfg(feature = "gpu")]
+            InternalServerKey::Cuda(_) => {
+                panic!("Cuda devices do not support trailing_ones yet");
+            }
+        })
+    }
+
+    /// Returns the base 2 logarithm of the number, rounded down.
+    ///
+    /// Result has no meaning if self encrypts 0. See [Self::checked_ilog2]
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tfhe::prelude::*;
+    /// use tfhe::{generate_keys, set_server_key, ConfigBuilder, FheBool, FheUint16};
+    ///
+    /// let (client_key, server_key) = generate_keys(ConfigBuilder::default());
+    /// set_server_key(server_key);
+    ///
+    /// let a = FheUint16::encrypt(2u16, &client_key);
+    ///
+    /// let result = a.ilog2();
+    /// let decrypted: u16 = result.decrypt(&client_key);
```

Comment:
we can decrypt to a smaller data type ?

Reply:
Yes because the function is implemented generically.

Haven't had the time to try an impl based on an associated `Clear` type that would be tied to each FheUint

Reply:
ok